src/m3ninx/index/segment/fst/encoding/docs/compressed_data.go

```diff
@@ -0,0 +1,170 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package docs
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3/src/m3ninx/doc"
+	"github.com/m3db/m3/src/m3ninx/index/segment/fst/encoding"
+)
+
+// compressedDataWriter writes the data file for documents.
+type compressedDataWriter struct {
+	writer        io.Writer
+	accumulator   *encoding.Encoder
+	docEncoder    *Encoder
+	compressor    Compressor
+	currentOffset offset
+
+	copts CompressionOptions
+}
+
+// newCompressedDataWriter returns a new compressedDataWriter.
+func newCompressedDataWriter(co CompressionOptions, w io.Writer) (dataWriter, error) {
+	c, err := co.newCompressor()
+	if err != nil {
+		return nil, err
+	}
+	return &compressedDataWriter{
+		writer:      w,
+		accumulator: encoding.NewEncoder(co.PageSize),
+		docEncoder:  NewEncoder(co.PageSize),
+		compressor:  c,
+		copts:       co,
+	}, nil
+}
+
+func (w *compressedDataWriter) Write(d doc.Document) (offset, error) {
+	docBytes, err := w.docEncoder.Encode(d)
+	if err != nil {
+		return offset{}, err
+	}
+
+	// check if adding this document overflows the page.
+	if w.currentOffset.docOffset+uint64(len(docBytes)) > uint64(w.copts.PageSize) {
+		// flush accumulated docs.
+		if err := w.compressAndFlush(); err != nil {
+			return offset{}, err
+		}
+	}
+
+	writeOffset := w.currentOffset
+	w.accumulator.PutRawBytes(docBytes)
+	w.currentOffset.docOffset += uint64(len(docBytes))
+	return writeOffset, nil
+}
+
+func (w *compressedDataWriter) compressAndFlush() error {
+	// compress the accumulated bytes.
+	original := w.accumulator.Bytes()
+	originalSize := len(original)
+	if originalSize == 0 {
+		return nil
+	}
+
+	compressed, err := w.compressor.Compress(original)
+	if err != nil {
+		return err
+	}
+
+	// write the compressed bytes to the writer.
+	w.accumulator.Reset()
+	w.accumulator.PutBytes(compressed)
+	bytes := w.accumulator.Bytes()
+
+	if _, err := w.writer.Write(bytes); err != nil {
+		return err
+	}
+
+	// update offsets and reset accumulator.
+	w.currentOffset.pageOffset += uint64(len(bytes))
+	w.currentOffset.docOffset = 0
+	w.accumulator.Reset()
+	return nil
+}
+
+// Flush flushes the compressedDataWriter.
+func (w *compressedDataWriter) Flush() error {
+	return w.compressAndFlush()
+}
+
+// Reset resets the compressedDataWriter.
+func (w *compressedDataWriter) Reset(wr io.Writer) {
+	w.writer = wr
+	w.compressor.Reset()
+	w.accumulator.Reset()
+	w.docEncoder.Reset()
+	w.currentOffset = offset{}
+}
+
+// compressedDataReader is a reader for the data file for documents.
+type compressedDataReader struct {
+	data  []byte
+	ropts ReaderOptions
+}
+
+// newCompressedDataReader returns a new compressedDataReader.
+func newCompressedDataReader(ropts ReaderOptions, data []byte) (dataReader, error) {
```

Comment:
Agreed but definitely will be hard to iterate the FST without using mmap, just because of how vellum works right now.  Guess we could totally refactor the read side of the library at some point..

Reply:
Actually, this is the documents data which we could definitely avoid using mmap for.  Is it worth thinking about removing mmap for the documents data as we introduce compression to just these files anyhow?